src/allocationDispatcher.c

```diff
@@ -0,0 +1,55 @@
+#include "allocationDispatcher.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <assert.h>
+
+ALLOCATION_DISPATCHER* adInitialize(size_t blockSizes[], size_t numBlocks[], size_t numAllocators, void* memoryPool){

```

Comment:
I can barely understand what this code is doing. Why is `memoryHead` being assigned multiple times? What does the variable name mean? If it has a fixed memory layout, then it should be cast to an appropriately defined struct expressing as much.

Reply:
this is kind of difficult, as the memory layout isn't quite fixed - each underlying allocator can be a different size. Added a comment explaining what is going on.

Reply:
Then this is our chance to pick a sensible memory layout.

I'm even not really sure it makes sense to make this as configurable as suggested by this interface. For instance, it makes no sense initialize the the allocation pool with sizes `{1, 2, 7, 3, 5, 49}` or something crazy like that. I'm not even convinced there's much benefit in configuring this differently at runtime.

@sedwards-lab do you have any thoughts on this? Are there applications where we would dramatically benefit from having different block size configurations?

Reply:
Also, we really should be using an on-demand paging mechanism, so that the underlying allocator can request new pages dynamically. We should also fix the page size to make things easier, and maybe just have allocator pools for blocks of size 4, 16, 64, 256, and 1024 (or something like that). That would suggest an interface like:

```
allocation_dispatcher_t *ad_initialize(/* other config */ void *(*get_page)(size_t));
```

where `get_page` is a callback that the dispatcher can use to request more pages to expand its memory pool; the `size_t` parameter would be used to indicate how many pages worth of memory are required. For instance, on UNIX-like systems, `get_page` could use the `mmap` syscall to feed more pages to each allocator.